File: Implementations C++/Dynamic Array/dynamic_array.cpp
#include <bits/stdc++.h>

using namespace std;

template<class genType, int size = 50>
class dynamic_array {
    public:
        dynamic_array() {
            elements = new genType[size];
        }
        void push_back(genType element) {
            capacity++;
            if (capacity == total_size - 1) {
                genType* new_array = new genType[total_size * 2];
                for (int i = 0; i < capacity - 1; i++) {
                    new_array[i] = elements[i];
                }
                new_array[capacity - 1] = element;
                delete [] elements;
                elements = new_array;
                total_size = total_size * 2;
            } else {
                elements[capacity - 1] = element;
            }
        }
        genType pop_back() {
            if (capacity == 0) {
                return elements[capacity];
            }
            capacity--;
            return elements[capacity];
        }

        int get_capacity() {
            return capacity;
        }

        int get_total_size() {
            return total_size;
        }

        void print_elements() {
            if (capacity == 0)
                return;
            for (int i = 0; i < capacity; i++) {
                cout << to_string(elements[i]) << ' ';
            }
            cout << '\n';
        }
    private:
        int capacity = 0;
        int total_size = size;
        genType* elements;
};

int main() {
    dynamic_array<int, 10> d;
    for (int i = 0; i < 11; i++) {
        d.push_back(i + 1);
    }
    d.print_elements();
    cout << d.get_capacity() << '\n';
    cout << d.get_total_size() << '\n';
    for (int i = 0; i < 11; i++) {
        cout << d.pop_back() << '\n';
    }
    d.print_elements();
    cout << d.get_capacity() << '\n';
    cout << d.get_total_size() << '\n';
    return 0;
}

File: Implementations Java/src/com/dynamic_array/DynamicArray.java
package com.dynamic_array;

public class DynamicArray<T> {
    private int size;
    private int capacity;
    private T[] elements;

    public DynamicArray(int size) {
        this.size = size;
        this.capacity++;
        this.elements = (T[]) new Object[size];
    }

    public void pushBack(T element) {
        this.capacity++;
        if (this.capacity == this.size - 1) {
            T[] newArray = (T[]) new Object[this.size * 2];
            for (int i = 0; i < capacity - 1; i++) {
                newArray[i] = this.elements[i];
            }
            newArray[capacity - 1] = element;
            this.elements = newArray;
            this.size = this.size * 2;
        } else {
            elements[this.capacity - 1] = element;
        }
    }

    public T popBack() {
        if (capacity == 0) {
            return null;
        }
        capacity--;
        return elements[this.capacity];
    }

    public int getCapacity() {
        return this.capacity - 1;
    }

    public int getSize() {
        return this.size;
    }

    @Override
    public String toString() {
        if (this.capacity == 1)
            return "";
        String res = "[";
        for (int i = 0; i < this.capacity - 1; i++) {
            if (this.elements[i] != null)
                res += this.elements[i] + ", ";
        }
        res += this.elements[this.capacity - 1] + "]";
        return res;
    }
}
